db/db_impl/db_impl_merge.cc

```diff
@@ -0,0 +1,390 @@
+  cf_db_super_versions.resize(num_cfs);
+  for (size_t cf_i = 0; cf_i < num_cfs; cf_i++) {
+    cf_db_super_versions[cf_i].resize(db_impls.size());
+    auto* this_cfd = this_cfds[cf_i];
+    auto& cf_name = this_cfd->GetName();
+    autovector<MemTable*> mems;
+    for (size_t db_i = 0; db_i < db_impls.size(); db_i++) {
+      auto& db = db_impls[db_i];
+      auto cfd = db->versions_->GetColumnFamilySet()->GetColumnFamily(cf_name);
+      if (cfd == nullptr || cfd->IsDropped()) {
+        cf_db_super_versions[cf_i][db_i] = nullptr;
+        continue;
+      }
+
+      if (merge_options.merge_memtable) {
+        if (!cfd->mem()->IsEmpty()) {
+          // Freeze memtable. Only immutable memtables can be shared.
+          auto& cf_mopts = *cfd->GetLatestMutableCFOptions();
+          auto* new_mem = cfd->ConstructNewMemtable(cf_mopts, max_seq_number);
+          new_mem->Ref();
+          cfd->imm()->Add(cfd->mem(), &to_delete);
+#ifndef ROCKSDB_LITE
+          MemTableInfo memtable_info;
+          memtable_info.cf_name = cf_name;
+          memtable_info.first_seqno = cfd->mem()->GetFirstSequenceNumber();
+          memtable_info.earliest_seqno =
+              cfd->mem()->GetEarliestSequenceNumber();
+          memtable_info.num_entries = cfd->mem()->num_entries();
+          memtable_info.num_deletes = cfd->mem()->num_deletes();
+          db->NotifyOnMemTableSealed(cfd, memtable_info);
+#endif  // ROCKSDB_LITE
+          cfd->SetMemtable(new_mem);
+          SuperVersionContext sv_context(/* create_superversion */ true);
+          db->InstallSuperVersionAndScheduleWork(cfd, &sv_context, cf_mopts);
+          sv_context.Clean();
+        }
+        assert(cfd->mem()->IsEmpty());
+
+        // [B] Bump log number for active memtable. Even though it's not
+        // shared, it must still be larger than other shared immutable
+        // memtables.
+        cfd->mem()->SetNextLogNumber(max_log_number);
+        cfd->imm()->ExportMemtables(&mems);
+      }
+
+      // Acquire super version.
+      cf_db_super_versions[cf_i][db_i] = cfd->GetSuperVersion()->Ref();
+    }
+    for (auto mem : mems) {
+      assert(mem != nullptr);
+      mem->Ref();
+      // [B] Bump log number for shared memtables.
+      mem->SetNextLogNumber(max_log_number);
+      this_cfd->imm()->Add(mem, &to_delete);
+    }
+    this_cfd->mem()->SetNextLogNumber(max_log_number);
+  }
+  for (size_t i = 0; i < all_db_impls.size(); i++) {
+    auto* db = all_db_impls[i];
+    bool check_log_number = (i == 0 || merge_options.allow_source_write) &&
+                            merge_options.merge_memtable;
+    if (check_log_number && max_log_number != db->logfile_number_) {
+      assert(max_log_number > db->logfile_number_);
+      // [B] Create a new WAL so that future memtable will use the correct log
+      // number as well.
+      log::Writer* new_log = nullptr;
+      s = db->CreateWAL(max_log_number, 0 /*recycle_log_number*/,
```

Comment:
what if wal is disabled?

Reply:
It doesn't matter. WAL will be created regardless of whether user writes with disableWAL=true.